### PR TITLE
Throw for invalid TryParse and BindAsync

### LINKED ISSUE #36523: Throw for invalid TryParse and BindAsync methods
If your `TryParse` method has a nullable out parameter and is a struct it fails to find the TryParse method and will throw:
`System.InvalidOperationException : No public static bool MyBindAsyncRecord.TryParse(string, out MyBindAsyncRecord) method found for myBindAsyncRecord.`

```csharp
private record struct MyBindAsyncRecord(Uri Uri)
{
    public static bool TryParse(string? value, out MyBindAsyncRecord? result) =>
        throw new NotImplementedException();
}
```

This is because the method parameter type is ```Nullable`1[[MyBindAsyncRecord]]&``` and we pass `MyBindAsyncRecord&`
https://github.com/dotnet/aspnetcore/blob/de277154511effa33f946a34ccc77deb074ce1c2/src/Shared/ParameterBindingMethodCache.cs#L107

### PR DESCRIPTION
**Pull request by BrennanConroy**
Fixes https://github.com/dotnet/aspnetcore/issues/36523

Tried to make the error semi-nice, there is room for improvement if we're willing to loop over the args and format them

**BindAsync**
```
BindAsync method found on InvalidWrongReturnBindAsyncStruct with incorrect format. Must be a static method with format
ValueTask<InvalidWrongReturnBindAsyncStruct> BindAsync(HttpContext context, ParameterInfo parameter)
ValueTask<InvalidWrongReturnBindAsyncStruct> BindAsync(HttpContext context)
ValueTask<Nullable<InvalidWrongReturnBindAsyncStruct>> BindAsync(HttpContext context, ParameterInfo parameter)
ValueTask<Nullable<InvalidWrongReturnBindAsyncStruct>> BindAsync(HttpContext context)
but found
static System.Threading.Tasks.Task`1[Microsoft.AspNetCore.Http.Extensions.Tests.ParameterBindingMethodCacheTests+InvalidWrongReturnBindAsyncStruct] BindAsync(Microsoft.AspNetCore.Http.HttpContext, System.Reflection.ParameterInfo)
```
**TryParse**
```
TryParse method found on InvalidNonStaticTryParseClass with incorrect format. Must be a static method with format
bool TryParse(string, IFormatProvider, out InvalidNonStaticTryParseClass)
bool TryParse(string, out InvalidNonStaticTryParseClass)
but found
not-static Boolean TryParse(System.String, System.IFormatProvider, InvalidNonStaticTryParseClass ByRef)
```

### CODE REVIEW

File: src/Http/Http.Extensions/test/ParameterBindingMethodCacheTests.cs [hunk 1]
```diff
@@ -274,6 +274,63 @@ public void FindBindAsyncMethod_FindsNonNullableReturningBindAsyncMethodGivenNul
...
+        [InlineData(typeof(InvalidWrongReturnBindAsyncClass))]
+        [InlineData(typeof(InvalidWrongParamBindAsyncStruct))]
+        [InlineData(typeof(InvalidWrongParamBindAsyncClass))]
+        public void FindBindAsyncMethod_ThrowsIfInvalidBindAsyncOnType(Type type)
+        {
+            var cache = new ParameterBindingMethodCache();
+            var parameter = new MockParameterInfo(type, "anything");
+            var ex = Assert.Throws<InvalidOperationException>(
+                () => cache.FindBindAsyncMethod(parameter));
+            Assert.StartsWith($"BindAsync method found on {type.Name} with incorrect format. Must be a static method with format", ex.Message);
```
Comment (halter73) [inline src/Http/Http.Extensions/test/ParameterBindingMethodCacheTests.cs:316]: How does this look with generic types? Do we want to use TypeNameHelper?

Comment (BrennanConroy) [reply src/Http/Http.Extensions/test/ParameterBindingMethodCacheTests.cs]: Changed to use TypeNameHelper

**Review by halter73 — APPROVED** [review]

File: src/Shared/ParameterBindingMethodCache.cs [hunk 2]
```diff
@@ -117,11 +119,24 @@ public bool HasTryParseMethod(ParameterInfo parameter)
...
-                if (methodInfo != null)
+                if (methodInfo is not null && methodInfo.ReturnType == typeof(bool))
                 {
                     return (expression) => Expression.Call(methodInfo, TempSourceStringExpr, expression);
                 }
 
+                if (type.GetMethod("TryParse", BindingFlags.Public | BindingFlags.Static | BindingFlags.Instance) is MethodInfo invalidMethod)
+                {
+                    var stringBuilder = new StringBuilder();
+                    stringBuilder.AppendLine(CultureInfo.InvariantCulture, $"TryParse method found on {TypeNameHelper.GetTypeDisplayName(type)} with incorrect format. Must be a static method with format");
```
Comment (halter73) [inline src/Shared/ParameterBindingMethodCache.cs:130]: I think we should probably use `fullName: false` like we do elsewhere in RequestDelegateFactory.

**Pull request merged**